src/analytics.js

```diff
@@ -0,0 +1,201 @@
+};
+
+const isCI = () => {
+  const ci = guessCI();
+  return ci.length > 0;
+};
+
+const groupId = async () => {
+  if (!isCI()) return '';
+
+  const ci = guessCI();
+  let rawId = 'CI';
+  if (ci === 'github') {
+    rawId = `${GITHUB_SERVER_URL}/${GITHUB_REPOSITORY_OWNER}`;
+  } else if (ci === 'gitlab') {
+    rawId = `${CI_SERVER_URL}/${CI_PROJECT_ROOT_NAMESPACE}`;
+  } else if (ci === 'bitbucket') {
+    rawId = BITBUCKET_WORKSPACE;
+  }
+
+  return await deterministic(rawId);
+};
+
+const userId = async ({ cml } = {}) => {
+  try {
+    if (isCI()) {
+      let rawId;
+      const ci = guessCI();
+      if (ci === 'github') {
+        const { name, login, id } = await cml
+          .getDriver()
+          .user({ name: GITHUB_ACTOR });
+        rawId = `${name || ''} ${login} ${id}`;
+      } else if (ci === 'gitlab') {
+        rawId = `${GITLAB_USER_NAME} ${GITLAB_USER_LOGIN} ${GITLAB_USER_ID}`;
+      } else if (ci === 'bitbucket') {
+        rawId = BITBUCKET_STEP_TRIGGERER_UUID;
+      } else {
+        rawId = await exec(`git log -1 --pretty=format:'%ae'`);
+      }
+
+      return await deterministic(rawId);
+    }
+
+    let id = uuidv4();
+    const oldPath = userConfigDir('dvc/user_id', 'iterative');
+    const newPath = userConfigDir('iterative/telemetry');
+    const readId = async ({ fpath }) => {
+      const json = (await fs.readFile(fpath)).toString('utf-8');
+      const { user_id: id } = JSON.parse(json);
+
+      return id;
+    };
+
+    const writeId = async ({ fpath, id }) => {
+      await fs.mkdir(path.dirname(fpath), { recursive: true });
+      await fs.writeFile(fpath, JSON.stringify({ user_id: id }));
+    };
+
+    if (await fileExists(newPath)) {
+      id = await readId({ fpath: newPath });
+    } else {
+      if (await fileExists(oldPath)) id = await readId({ fpath: oldPath });
+
+      await writeId({ fpath: newPath, id });
+    }
+
+    if (!(await fileExists(oldPath)) && id !== ID_DO_NOT_TRACK)
+      await writeId({ fpath: oldPath, id });
+
+    return id;
+  } catch (err) {
+    winston.debug(`userId failure: ${err.message}`);
+  }
+};
+
+const OS = () => {
+  if (process.platform === 'win32') return 'windows';
+  if (process.platform === 'darwin') return 'macos';
+
+  return process.platform;
+};
+
+const jitsuEventPayload = async ({
+  action = '',
+  error = '',
+  extra = {},
+  cml
+} = {}) => {
+  try {
+    const { cloud: backend = '', ...extraRest } = extra;
+    extraRest.ci = guessCI();
+
+    return {
+      user_id: await userId({ cml }),
+      group_id: await groupId(),
+      action,
+      interface: 'cli',
+      tool_name: 'cml',
+      tool_version: VERSION,
+      tool_source: '',
+      os_name: OS(),
+      os_version: os.release(),
+      backend,
+      error,
+      extra: extraRest
+    };
+  } catch (err) {
+    return {};
+  }
+};
+
+const send = async ({
+  event,
+  endpoint = ITERATIVE_ANALYTICS_ENDPOINT,
+  token = ITERATIVE_ANALYTICS_TOKEN
+} = {}) => {
+  try {
```

Comment:
```suggestion
  await require('./utils').waitForever();
  try {
```

Reply:
As it is right now? the code will wait until we send the analytics

Reply:
How can I test it?

Reply:
```js
export ITERATIVE_ANALYTICS_ENDPOINT=http://localhost:3004
export ITERATIVE_ANALYTICS_TOKEN=XXXX
cml publish assets/logo.png
```

Reply:
my own jitsu

Reply:
you can setup yours or go against the real one or use an echo server

Reply:
But it — obviously — doesn't hang forever. Thus, we can't prove that the `analytics.send` function will return before exiting.

Reply:
?

Reply:
Short and sweet: the process exits regardless of `analytics.send` being still in progress; if just works because server responses are fast enough.

Reply:
Resolving after some out-of-band clarifications: it turns out that I didn't read the latest changes and was commenting while thinking of an outdated version.